components/script/dom/htmlinputelement.rs

```diff
@@ -2532,3 +2895,67 @@ fn milliseconds_to_datetime(value: f64) -> Result<NaiveDateTime, ()> {
     let nanoseconds = milliseconds * 1e6;
     NaiveDateTime::from_timestamp_opt(seconds as i64, nanoseconds as u32).ok_or(())
 }
+
+// This is used to compile JS-compatible regex provided in pattern attribute
+// that matches only the entirety of string.
+// https://html.spec.whatwg.org/multipage/#compiled-pattern-regular-expression
+fn compile_pattern(cx: SafeJSContext, pattern_str: &str, out_regex: MutableHandleObject) -> bool {
+    // First check if pattern compiles...
+    if new_js_regex(cx, pattern_str, out_regex) {
+        // ...and if it does make pattern that matches only the entirety of string
+        let pattern_str = format!("^(?:{})$", pattern_str);
+        new_js_regex(cx, &pattern_str, out_regex)
+    } else {
+        false
+    }
+}
+
+#[allow(unsafe_code)]
+fn new_js_regex(cx: SafeJSContext, pattern: &str, mut out_regex: MutableHandleObject) -> bool {
+    // TODO: This flag should be in mozjs, https://github.com/servo/mozjs/pull/224
+    const JSREG_UNICODE: c_uint = 0x10;
+    let pattern: Vec<u16> = pattern.encode_utf16().collect();
+    unsafe {
+        out_regex.set(JS_NewUCRegExpObject(
+            *cx,
+            pattern.as_ptr(),
+            pattern.len(),
+            JSREG_UNICODE,
+        ));
+        if out_regex.is_null() {
+            JS_ClearPendingException(*cx);
+            return false;
+        }
+    }
+    true
+}
+
+#[allow(unsafe_code)]
+fn matches_js_regex(cx: SafeJSContext, regex_object: HandleObject, value: &str) -> bool {
+    let mut value: Vec<u16> = value.encode_utf16().collect();
+
+    unsafe {
+        let mut is_regex = false;
+        assert!(JS_ObjectIsRegExp(*cx, regex_object, &mut is_regex));
+        assert!(is_regex);
+
+        rooted!(in(*cx) let mut rval = UndefinedValue());
+        let mut index = 0;
+
+        let ok = JS_ExecuteRegExpNoStatics(
+            *cx,
+            regex_object,
+            value.as_mut_ptr(),
+            value.len(),
+            &mut index,
+            true,
+            rval.handle_mut(),
+        );
+
+        if !ok {
+            JS_ClearPendingException(*cx);
+        }
+
+        !ok || !rval.is_null()
```

Comment:
Don't we want `ok && !rval.is_null()`? Am I misunderstanding what JS_ExecuteRegExpNoStatics returns?

Reply:
It returns false on error. If pattern doesn't match it returns true and set rval to null:

https://github.com/servo/mozjs/blob/b2f83932fe9d361face14efd03f2465b9262e687/mozjs/js/src/builtin/RegExp.cpp#L167-L177

I guess errors should be treated the same way as invalid patterns, i.e. element shouldn't suffer from pattern mismatch.

Reply:
I changed it to return `Result<bool, ()>` instead of `bool` to make it less confusing and errors are ignored in `suffers_from_pattern_mismatch`.